python/context.cpp

```diff
@@ -33,94 +34,81 @@ std::ostream& operator<<(std::ostream& o, const context_shim& ctx) {
           << ">";
 }
 
-// A Python shim that holds the information that describes an arb::proc_allocation.
-struct proc_allocation_shim {
-    std::optional<int> gpu_id = {};
-    unsigned long num_threads = 1;
-    bool bind_procs = false;
-    bool bind_threads = false;
-
-    proc_allocation_shim(unsigned threads, pybind11::object gpu, bool bp, bool bt) {
-        set_num_threads(threads);
-        set_gpu_id(gpu);
-        set_bind_procs(bp);
-        set_bind_threads(bt);
-    }
-
-    proc_allocation_shim(): proc_allocation_shim(1, pybind11::none(), false, false) {}
-
-    // getter and setter (in order to assert when being set)
-    void set_gpu_id(pybind11::object gpu) {
-        gpu_id = py2optional<int>(gpu, "gpu_id must be None, or a non-negative integer", is_nonneg());
-    };
-
-    void set_num_threads(unsigned threads) {
-        if (0==threads) {
-            throw arb::zero_thread_requested_error(threads);
-        }
-        num_threads = threads;
-    };
-
-    bool get_bind_threads() const { return bind_threads; }
-    bool get_bind_procs() const { return bind_procs; }
-
-    void set_bind_threads(bool f) { bind_threads = f; }
-    void set_bind_procs(bool f) { bind_procs = f; }
+std::ostream& operator<<(std::ostream& o, const proc_allocation_shim& alloc) {
+    return o << "<arbor.proc_allocation: threads " << alloc.get_num_threads() <<
+                ", gpu_id " << util::to_string(alloc.get_gpu_id()) << 
+                ", bind_threads " << util::to_string(alloc.get_bind_threads()) << 
+                ", bind_procs " << util::to_string(alloc.get_bind_procs()) << 
+                ">";
+}
 
+// proc_alloc getter and setter (in order to assert when being set)
+proc_allocation_shim::proc_allocation_shim(unsigned threads, pybind11::object gpu, bool bp, bool bt) {
+    set_num_threads(threads);
+    set_gpu_id(gpu);
+    set_bind_procs(bp);
+    set_bind_threads(bt);
+}
 
-    std::optional<int> get_gpu_id() const { return gpu_id; }
-    unsigned get_num_threads() const { return num_threads; }
-    bool has_gpu() const { return bool(gpu_id); }
+void proc_allocation_shim::set_gpu_id(pybind11::object gpu) {
+    auto gpu_id = py2optional<int>(gpu, "gpu_id must be None, or a non-negative integer", is_nonneg());
+    proc_allocation.gpu_id = gpu_id.value_or(-1);
+};
 
-    // helper function to use arb::make_context(arb::proc_allocation)
-    arb::proc_allocation allocation() const {
-        return arb::proc_allocation(num_threads, gpu_id.value_or(-1), bind_procs, bind_threads);
+void proc_allocation_shim::set_num_threads(unsigned threads) {
+    if (0==threads) {
```

Comment:
Don't we throw this from C++?

Reply:
`arb:::proc_allocation` does not to any checking. Nor does anything else for that matter. Only the pyobject ever did.